examples/llm_plus_gnn/g_retriever.py

```diff
@@ -0,0 +1,472 @@
+    print(f'Precision: {precision:.4f}')
+    print(f'Recall: {recall:.4f}')
+    print(f'F1: {f1:.4f}')
+
+    return hit
+
+
+def save_params_dict(model, save_path):
+    state_dict = model.state_dict()
+    param_grad_dict = {
+        k: v.requires_grad
+        for (k, v) in model.named_parameters()
+    }
+    for k in list(state_dict.keys()):
+        if k in param_grad_dict.keys() and not param_grad_dict[k]:
+            # delete parameters that do not require gradient
+            del state_dict[k]
+    torch.save(state_dict, save_path)
+
+
+def load_params_dict(model, save_path):
+    state_dict = torch.load(save_path)
+    model.load_state_dict(state_dict)
+    return model
+
+
+def get_loss(model, batch, model_save_name) -> torch.Tensor:
+    if model_save_name == "llm":
+        return model(batch.question, batch.label, batch.desc)
+    else:
+        return model(batch.question, batch.x, batch.edge_index, batch.batch,
+                     batch.ptr, batch.label, batch.edge_attr, batch.desc)
+
+
+def inference_step(model, batch, model_save_name):
+    if model_save_name == "llm":
+        return model.inference(batch.question, batch.desc)
+    else:
+        return model.inference(batch.question, batch.x, batch.edge_index,
+                               batch.batch, batch.ptr, batch.edge_attr,
+                               batch.desc)
+
+
+def train(since, num_epochs, hidden_channels, num_gnn_layers, batch_size,
+          eval_batch_size, lr, loss_fn, inference_fn, model=None, dataset=None,
+          checkpointing=False, tiny_llama=False):
+    def adjust_learning_rate(param_group, LR, epoch):
+        # Decay the learning rate with half-cycle cosine after warmup
+        min_lr = 5e-6
+        warmup_epochs = 1
+        if epoch < warmup_epochs:
+            lr = LR
+        else:
+            lr = min_lr + (LR - min_lr) * 0.5 * (
+                1.0 + math.cos(math.pi * (epoch - warmup_epochs) /
+                               (num_epochs - warmup_epochs)))
+        param_group["lr"] = lr
+        return lr
+
+    seed_everything(42)
+    if dataset is None:
+        dataset = WebQSPDataset()
+        gc.collect()
+    idx_split = dataset.split_idxs
+
+    # Step 1: Build Node Classification Dataset
+    train_dataset = [dataset[i] for i in idx_split['train']]
+    val_dataset = [dataset[i] for i in idx_split['val']]
+    test_dataset = [dataset[i] for i in idx_split['test']]
+
+    train_loader = DataLoader(train_dataset, batch_size=batch_size,
+                              drop_last=True, pin_memory=True, shuffle=True)
+    val_loader = DataLoader(val_dataset, batch_size=eval_batch_size,
+                            drop_last=False, pin_memory=True, shuffle=False)
+    test_loader = DataLoader(test_dataset, batch_size=eval_batch_size,
+                             drop_last=False, pin_memory=True, shuffle=False)
+
+    # Step 2: Build Model
+    if model is None:
+        gc.collect()
+        if tiny_llama:
```

Comment:
Can we define an argument in argparse such as `model`. We can check if it's tinyllama and set corresponding arguments here.

Reply:
the issue with doing something like that is there are too many knobs such mlp out_dim and num_llm_params which must vary with the LLM used. We would need a user to set correctly through argparser for each  model. This code serves as an example for which they can look into the api and adjust the dimensions of their model as needed for custom workflows.